x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiUnifiedStreamingProcessor.java

```diff
@@ -0,0 +1,302 @@
+import org.elasticsearch.xpack.inference.external.response.streaming.ServerSentEvent;
+import org.elasticsearch.xpack.inference.external.response.streaming.ServerSentEventField;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.elasticsearch.xpack.inference.external.response.XContentUtils.moveToFirstToken;
+
+public class OpenAiUnifiedStreamingProcessor extends DelegatingProcessor<Deque<ServerSentEvent>, ChunkedToXContent> {
+    public static final String FUNCTION_FIELD = "function";
+    private static final Logger logger = LogManager.getLogger(OpenAiUnifiedStreamingProcessor.class);
+    private static final String FAILED_TO_FIND_FIELD_TEMPLATE = "Failed to find required field [%s] in OpenAI chat completions response";
+
+    private static final String CHOICES_FIELD = "choices";
+    private static final String DELTA_FIELD = "delta";
+    private static final String CONTENT_FIELD = "content";
+    private static final String DONE_MESSAGE = "[done]";
+    private static final String REFUSAL_FIELD = "refusal";
+    private static final String TOOL_CALLS_FIELD = "tool_calls";
+    public static final String ROLE_FIELD = "role";
+    public static final String FINISH_REASON_FIELD = "finish_reason";
+    public static final String INDEX_FIELD = "index";
+    public static final String OBJECT_FIELD = "object";
+    public static final String MODEL_FIELD = "model";
+    public static final String ID_FIELD = "id";
+    public static final String CHOICE_FIELD = "choice";
+    public static final String USAGE_FIELD = "usage";
+    public static final String TYPE_FIELD = "type";
+    public static final String NAME_FIELD = "name";
+    public static final String ARGUMENTS_FIELD = "arguments";
+    public static final String COMPLETION_TOKENS_FIELD = "completion_tokens";
+    public static final String PROMPT_TOKENS_FIELD = "prompt_tokens";
+    public static final String TOTAL_TOKENS_FIELD = "total_tokens";
+
+    private final Deque<StreamingUnifiedChatCompletionResults.ChatCompletionChunk> buffer = new LinkedBlockingDeque<>();
+
+    @Override
+    protected void onRequest(long n) {
+        if (buffer.isEmpty()) {
+            super.onRequest(n);
+        } else {
+            downstream().onNext(new StreamingUnifiedChatCompletionResults.Results(singleItem(buffer.poll())));
+        }
+    }
+
+    @Override
+    protected void next(Deque<ServerSentEvent> item) throws Exception {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        var results = new ArrayDeque<StreamingUnifiedChatCompletionResults.ChatCompletionChunk>(item.size());
```

Comment:
Follow up, we should refactor this class and OpenAiStreamingProcessor since they're nearly identical. The only difference is how the result is reported. Issue to track: https://github.com/elastic/elasticsearch/issues/118107